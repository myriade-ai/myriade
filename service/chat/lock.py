from threading import Lock

from flask_socketio import emit

conversation_stop_flags: dict[str, bool] = {}
stop_flag_lock = Lock()


class StopException(Exception):
    pass


class STATUS:
    RUNNING = "running"
    CLEAR = "clear"
    TO_STOP = "to_stop"
    ERROR = "error"


def check_and_clear_stop_flag(conversation_id: str) -> bool:
    """Thread-safe check and clear of stop flag for a conversation.

    Returns:
        True if stop flag was set (and has been cleared), False otherwise
    """
    with stop_flag_lock:
        if conversation_stop_flags.get(conversation_id):
            del conversation_stop_flags[conversation_id]
            return True
        return False


def set_stop_flag(conversation_id: str) -> None:
    """Thread-safe setting of stop flag for a conversation."""
    with stop_flag_lock:
        conversation_stop_flags[conversation_id] = True


def clear_stop_flag(conversation_id: str) -> None:
    """Thread-safe clearing of stop flag for a conversation."""
    with stop_flag_lock:
        conversation_stop_flags.pop(conversation_id, None)


def emit_status(conversation_id, status, error=None):
    emit(
        "status",
        {
            "conversation_id": str(conversation_id),
            "status": status,
            "error": str(error),
        },
    )
