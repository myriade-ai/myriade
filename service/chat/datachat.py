import os

import nest_asyncio
import yaml
from autochat import Autochat, Message
from autochat.chat import StopLoopException

from back.models import Chart, Conversation, ConversationMessage, Query
from chat.dbt_utils import DBT
from chat.lock import STATUS, StopException, emit_status
from chat.notes import Notes
from chat.tools.database import DatabaseTool
from chat.tools.echarts import EchartsTool
from chat.tools.workspace import WorkspaceTool
from chat.utils import parse_answer_text

# Workaround because of eventlet doesn't support loop in loop ?
nest_asyncio.apply()

AUTOCHAT_PROVIDER = os.getenv("AUTOCHAT_PROVIDER", "openai")


class DatabaseChat:
    """
    Chatbot assistant with a database, execute functions.
    """

    def __init__(
        self,
        session,
        database_id,
        conversation_id=None,
        stop_flags=None,
        model=None,
        project_id=None,
        user_id=None,
    ):
        self.session = session
        if conversation_id is None:
            # Create conversation object
            self.conversation = self._create_conversation(
                databaseId=database_id, project_id=project_id, user_id=user_id
            )
        else:
            self.conversation = (
                self.session.query(Conversation).filter_by(id=conversation_id).first()
            )

        # Add a datalake object to the request
        self.datalake = self.conversation.database.create_datalake()
        if stop_flags is None:
            self.stop_flags = {}
        else:
            self.stop_flags = stop_flags
        self.model = model

        self.dbt = None
        if (
            self.conversation.database.dbt_catalog
            and self.conversation.database.dbt_manifest
        ):
            self.dbt = DBT(
                catalog=self.conversation.database.dbt_catalog,
                manifest=self.conversation.database.dbt_manifest,
            )

    def __del__(self):
        # On destruct, close the engine
        if hasattr(self, "datalake"):
            self.datalake.dispose()

    def _create_conversation(
        self, databaseId, name=None, project_id=None, user_id=None
    ):
        # Create conversation object
        conversation = Conversation(
            databaseId=databaseId,
            ownerId=user_id,
            name=name,
            projectId=project_id,
        )
        self.session.add(conversation)
        self.session.commit()
        return conversation

    def check_stop_flag(self):
        if self.stop_flags.get(str(self.conversation.id)):
            del self.stop_flags[str(self.conversation.id)]  # Remove the stop flag
            raise StopException("Query stopped by user")

    @property
    def context(self):
        context = {
            "DATABASE": {
                "name": self.conversation.database.name,
                "engine": self.conversation.database.engine,
            },
            "MEMORY": self.conversation.database.memory,
        }
        if self.conversation.project:
            context["PROJECT"] = {
                "name": self.conversation.project.name,
                "description": self.conversation.project.description,
                "tables": [
                    {"schema": table.schemaName, "table": table.tableName}
                    for table in self.conversation.project.tables
                ],
            }

        return yaml.dump(context)

    @property
    def chatbot(self):
        messages = [m.to_autochat_message() for m in self.conversation.messages]
        chatbot = Autochat.from_template(
            os.path.join(os.path.dirname(__file__), "..", "chat", "chat_template.txt"),
            provider=AUTOCHAT_PROVIDER,
            context=self.context,
            messages=messages,
        )

        chatbot.simple_response_callback = lambda response: Message(
            role="user",
            content="Only messages from 'answer' function call are\
            visible to the user. Use it to answer the user.",
        )

        chatbot.add_tool(
            DatabaseTool(self.session, self.conversation.database), "database"
        )
        chatbot.add_function(self.save_to_memory)
        chatbot.add_function(self.submit)
        chatbot.add_function(self.answer)
        chatbot.add_function(self.ask_user)
        chatbot.add_tool(WorkspaceTool(self.session, self.conversation.id), "workspace")
        chatbot.add_tool(
            EchartsTool(self.session, self.conversation.database),
            "echarts",
        )
        if self.dbt:
            chatbot.add_tool(self.dbt, self.conversation.database.name)
        if self.conversation.project:
            notes = Notes(self.session, chatbot, self.conversation.project)
            chatbot.add_tool(notes, "Notes")

        if self.model:
            chatbot.model = self.model

        return chatbot

    def save_to_memory(self, text: str):
        """
        Add a text to the AI's memory
        Args:
            text: The text to add to the memory
        """
        if self.conversation.database.memory is None:
            self.conversation.database.memory = text
        else:
            self.conversation.database.memory += "\n" + text
        self.session.commit()

    def submit(
        self,
        from_response: Message,
        queryId: int,
    ):
        """
        Give the final response from the user demand/query
        Args:
            queryId: The id of the query to execute
        """  # noqa: E501
        query = self.session.query(Query).filter_by(id=queryId).first()
        if not query:
            raise ValueError(f"Query with id {queryId} not found")
        # We update the message with the query id
        from_response.query_id = query.id
        from_response.isAnswer = True
        raise StopLoopException("We want to stop after submitting")

    def ask_user(self, question: str, from_response: Message):
        """
        Ask the user a question. Use it to ask for confirmation, for ambiguous queries,\
        etc.
        Use it only when it strictly necessary.
        Args:
            question: The question to ask the user
        """
        from_response.isAnswer = True  # TODO: is answer is not the correct termilogy
        raise StopLoopException("We want the user to answer")

    def answer(
        self,
        text: str,
        from_response: Message,
    ):
        """
        Give the final response from the user demand/query as a text.
        You can insert a query in the text using the <QUERY:{query_id}> tag.
        You can insert a chart in the text using the <CHART:{chart_id}> tag.
        Replace {query_id} and {chart_id} with the actual query id and chart id.
        You can only insert one query and one chart per message.
        Show the query only if the user asked for it.
        Show the chart only if the user asked for it or if that make sense to have it.
        """

        chunks = parse_answer_text(text)
        # Check query_id & chart_id
        for chunk in chunks:
            if chunk["type"] == "query":
                # extract the query id from the text
                query_id = chunk["query_id"]
                query = (
                    self.session.query(Query)
                    .join(ConversationMessage, Query.conversation_messages)
                    .filter(
                        Query.id == query_id,
                        ConversationMessage.conversationId == self.conversation.id,
                    )
                    .first()
                )
                if not query:
                    raise ValueError(
                        f"Query with id {query_id} not found in this conversation"
                    )

            if chunk["type"] == "chart":
                # extract the chart id from the text
                chart_id = chunk["chart_id"]
                # verify chart exists and belongs to this conversation
                chart = (
                    self.session.query(Chart)
                    .join(ConversationMessage, Chart.conversation_messages)
                    .filter(
                        Chart.id == chart_id,
                        ConversationMessage.conversationId == self.conversation.id,
                    )
                    .first()
                )
                if not chart:
                    raise ValueError(
                        f"Chart with id {chart_id} not found in this conversation"
                    )

        from_response.isAnswer = True
        raise StopLoopException("We want to stop after submitting")

    def _run_conversation(self):
        emit_status(self.conversation.id, STATUS.RUNNING)
        try:
            messages = [m.to_autochat_message() for m in self.conversation.messages]
            self.chatbot.load_messages(messages)
            for m in self.chatbot.run_conversation():
                self.check_stop_flag()
                # We re-emit the status in case the user has refreshed the page
                emit_status(self.conversation.id, STATUS.RUNNING)
                message = ConversationMessage.from_autochat_message(m)
                message.conversationId = self.conversation.id
                self.session.add(message)
                self.session.commit()
                yield message
            emit_status(self.conversation.id, STATUS.CLEAR)
        except Exception as e:
            emit_status(self.conversation.id, STATUS.ERROR, e)
            import traceback

            traceback.print_exc()
            raise

    def ask(self, question: str):
        if not self.conversation.name:
            self.conversation.name = question

        # If message is instance of string, then convert to ConversationMessage
        message = ConversationMessage(
            role="user",
            content=question,
            conversationId=self.conversation.id,
        )
        self.session.add(message)
        self.session.commit()
        yield message

        yield from self._run_conversation()
