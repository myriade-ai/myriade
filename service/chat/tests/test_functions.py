"""Use syrupy to test that the imported functions stay consistent.

We initialize the datachat class and test that the functions are consistent.
"""

import pytest
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker
from syrupy.assertion import SnapshotAssertion

from back.models import Base, Database, Project, User
from chat.datachat import DatabaseChat


@pytest.fixture
def session():
    """Create a new database session for testing"""
    # Create an in-memory SQLite database
    engine = create_engine("sqlite:///:memory:")

    # Create all tables
    Base.metadata.create_all(engine)

    # Create a sessionmaker
    TestingSessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)

    # Create a new session
    session = TestingSessionLocal()
    try:
        yield session
    finally:
        session.close()
        Base.metadata.drop_all(engine)


@pytest.fixture
def user(session):
    """Create a test user"""
    user = User(id="test-user", email="test@test.com")
    session.add(user)
    session.commit()
    return user


@pytest.fixture
def database(session):
    """Create a test database"""
    database = Database(
        name="test",
        _engine="sqlite",
        details={"filename": ":memory:"},
        memory="test memory",
        dbt_manifest={"sources": {}, "nodes": {}},
        dbt_catalog={"sources": {}, "nodes": {}},
    )
    session.add(database)
    session.commit()
    return database


@pytest.fixture
def project(session, database, user):
    """Create a test project"""
    project = Project(
        name="test",
        description="test project",
        tables=[],  # Add dummy tables if needed
        databaseId=database.id,
        creatorId=user.id,
    )
    session.add(project)
    session.commit()
    return project


@pytest.fixture
def datachat(session, database, project):
    """Create a datachat instance with a database and project"""
    return DatabaseChat(
        session=session,
        database_id=database.id,
        project_id=project.id,
        user_id="test-user",
    )


def test_functions(datachat, snapshot: SnapshotAssertion):
    """Test that the functions dictionary stays consistent"""
    chatbot = datachat.chatbot
    # Convert functions to a serializable format
    functions_list = [name for name, _ in chatbot.functions.items()]
    assert functions_list == snapshot


def test_functions_schema(datachat, snapshot: SnapshotAssertion):
    """Test that the functions schema stays consistent"""
    chatbot = datachat.chatbot
    assert chatbot.functions_schema == snapshot


def test_tools(datachat, snapshot: SnapshotAssertion):
    """Test that the tools dictionary stays consistent"""
    chatbot = datachat.chatbot
    # Convert tools to a serializable format
    tools_dict = {
        name: str(tool) if callable(tool) else tool.__class__.__name__
        for name, tool in chatbot.tools.items()
    }
    assert tools_dict == snapshot
