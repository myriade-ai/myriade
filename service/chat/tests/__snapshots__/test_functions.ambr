# serializer version: 1
# name: test_functions
  list([
    'DatabaseTool-database__save_to_memory',
    'DatabaseTool-database__sql_query',
    'think',
    'get_date',
    'submit',
    'answer',
    'ask_user',
    'EchartsTool-echarts__preview_render',
    'SemanticCatalog-semantic_catalog__create_entity',
    'SemanticCatalog-semantic_catalog__create_issue',
    'SemanticCatalog-semantic_catalog__delete_entity',
    'SemanticCatalog-semantic_catalog__read_issue',
    'SemanticCatalog-semantic_catalog__update_entity',
    'SemanticCatalog-semantic_catalog__update_issue',
    'DBT-dbt__fetch_model',
    'DBT-dbt__fetch_model_list',
    'DBT-dbt__search_models',
    'Notes-notes__close_note',
    'Notes-notes__create_note',
    'Notes-notes__delete_note',
    'Notes-notes__list_notes',
    'Notes-notes__open_note',
  ])
# ---
# name: test_functions_schema
  list([
    dict({
      'description': "Add text to the AI's memory",
      'name': 'DatabaseTool-database__save_to_memory',
      'parameters': dict({
        'properties': dict({
          'text': dict({
            'description': 'The text to add to the memory',
            'title': 'Text',
            'type': 'string',
          }),
        }),
        'required': list([
          'text',
        ]),
        'title': 'Input for `save_to_memory`',
        'type': 'object',
      }),
    }),
    dict({
      'description': 'Run an SQL query on the database and return the result',
      'name': 'DatabaseTool-database__sql_query',
      'parameters': dict({
        'properties': dict({
          'query': dict({
            'description': "The SQL query string to be executed. Don't forget to escape this if you use double quote.",
            'title': 'Query',
            'type': 'string',
          }),
          'title': dict({
            'default': '',
            'description': 'The name/title of the query',
            'title': 'Title',
            'type': 'string',
          }),
        }),
        'required': list([
          'query',
        ]),
        'title': 'Input for `sql_query`',
        'type': 'object',
      }),
    }),
    dict({
      'description': 'Think about the task at hand. It helps to reflect or decompose the situation.',
      'name': 'think',
      'parameters': dict({
        'properties': dict({
          'thought': dict({
            'description': 'A thought to think about.',
            'title': 'Thought',
            'type': 'string',
          }),
        }),
        'required': list([
          'thought',
        ]),
        'title': 'Input for `think`',
        'type': 'object',
      }),
    }),
    dict({
      'description': 'Get the current date as a string. Returns: the current date string in YYYY-MM-DD format',
      'name': 'get_date',
      'parameters': dict({
        'properties': dict({
        }),
        'title': 'Input for `get_date`',
        'type': 'object',
      }),
    }),
    dict({
      'description': 'Give the final response from the user demand/query',
      'name': 'submit',
      'parameters': dict({
        'properties': dict({
          'queryId': dict({
            'description': 'The id of the query to execute',
            'title': 'Queryid',
            'type': 'string',
          }),
        }),
        'required': list([
          'queryId',
        ]),
        'title': 'Input for `submit`',
        'type': 'object',
      }),
    }),
    dict({
      'description': "Give the final response from the user demand/query as a text. You can insert a query with it's preview result in the text using the <QUERY:{query_id}> tag. You can insert a chart in the text using the <CHART:{chart_id}> tag. Replace {query_id} and {chart_id} with the actual query id and chart id. Show the query / chart only if the user asked for it or if that make sense to have it.",
      'name': 'answer',
      'parameters': dict({
        'properties': dict({
          'text': dict({
            'title': 'Text',
            'type': 'string',
          }),
        }),
        'required': list([
          'text',
        ]),
        'title': 'Input for `answer`',
        'type': 'object',
      }),
    }),
    dict({
      'description': 'Ask the user a question. Use it to ask for confirmation, for ambiguous queries,    etc. Use it only when it strictly necessary.',
      'name': 'ask_user',
      'parameters': dict({
        'properties': dict({
          'question': dict({
            'description': 'The question to ask the user',
            'title': 'Question',
            'type': 'string',
          }),
        }),
        'required': list([
          'question',
        ]),
        'title': 'Input for `ask_user`',
        'type': 'object',
      }),
    }),
    dict({
      'description': 'Render a chart (using Echarts 4). This is not shown to the user, but this will create a chart object That you can reference in the answer response. --- Provide the chart_options without the "dataset" parameter We will fill the dataset.source automatically with the SQL result Don\'t forget to Map from Data to Charts (series.encode) using the correct names Don\'t use specific color in the chart_options unless the user asked for it When creating bar charts with ECharts, make sure to set the correct axis types.',
      'name': 'EchartsTool-echarts__preview_render',
      'parameters': dict({
        '$defs': dict({
          'Axis': dict({
            'properties': dict({
              'axisLabel': dict({
                'anyOf': list([
                  dict({
                    '$ref': '#/$defs/AxisLabel',
                  }),
                  dict({
                    'type': 'null',
                  }),
                ]),
                'default': None,
              }),
              'inverse': dict({
                'anyOf': list([
                  dict({
                    'type': 'boolean',
                  }),
                  dict({
                    'type': 'null',
                  }),
                ]),
                'default': None,
                'title': 'Inverse',
              }),
              'max': dict({
                'anyOf': list([
                  dict({
                    'type': 'integer',
                  }),
                  dict({
                    'type': 'number',
                  }),
                  dict({
                    'type': 'null',
                  }),
                ]),
                'default': None,
                'title': 'Max',
              }),
              'min': dict({
                'anyOf': list([
                  dict({
                    'type': 'integer',
                  }),
                  dict({
                    'type': 'number',
                  }),
                  dict({
                    'type': 'null',
                  }),
                ]),
                'default': None,
                'title': 'Min',
              }),
              'name': dict({
                'anyOf': list([
                  dict({
                    'type': 'string',
                  }),
                  dict({
                    'type': 'null',
                  }),
                ]),
                'default': None,
                'title': 'Name',
              }),
              'type': dict({
                'enum': list([
                  'value',
                  'category',
                  'time',
                  'log',
                ]),
                'title': 'Type',
                'type': 'string',
              }),
            }),
            'required': list([
              'type',
            ]),
            'title': 'Axis',
            'type': 'object',
          }),
          'AxisLabel': dict({
            'properties': dict({
              'formatter': dict({
                'anyOf': list([
                  dict({
                    'type': 'string',
                  }),
                  dict({
                    'type': 'null',
                  }),
                ]),
                'default': None,
                'title': 'Formatter',
              }),
              'interval': dict({
                'default': 0,
                'title': 'Interval',
                'type': 'integer',
              }),
              'rotate': dict({
                'anyOf': list([
                  dict({
                    'type': 'integer',
                  }),
                  dict({
                    'type': 'null',
                  }),
                ]),
                'default': None,
                'title': 'Rotate',
              }),
            }),
            'title': 'AxisLabel',
            'type': 'object',
          }),
          'ChartOptions': dict({
            'properties': dict({
              'dataZoom': dict({
                'anyOf': list([
                  dict({
                    'items': dict({
                      '$ref': '#/$defs/DataZoom',
                    }),
                    'type': 'array',
                  }),
                  dict({
                    'type': 'null',
                  }),
                ]),
                'default': None,
                'title': 'Datazoom',
              }),
              'grid': dict({
                'anyOf': list([
                  dict({
                    '$ref': '#/$defs/GridOptions',
                  }),
                  dict({
                    'type': 'null',
                  }),
                ]),
                'default': None,
              }),
              'series': dict({
                'items': dict({
                  '$ref': '#/$defs/Series',
                }),
                'title': 'Series',
                'type': 'array',
              }),
              'title': dict({
                'anyOf': list([
                  dict({
                    '$ref': '#/$defs/TitleOptions',
                  }),
                  dict({
                    'type': 'null',
                  }),
                ]),
                'default': None,
              }),
              'tooltip': dict({
                'anyOf': list([
                  dict({
                    '$ref': '#/$defs/Tooltip',
                  }),
                  dict({
                    'type': 'null',
                  }),
                ]),
                'default': None,
              }),
              'xAxis': dict({
                '$ref': '#/$defs/Axis',
              }),
              'yAxis': dict({
                '$ref': '#/$defs/Axis',
              }),
            }),
            'required': list([
              'xAxis',
              'yAxis',
              'series',
            ]),
            'title': 'ChartOptions',
            'type': 'object',
          }),
          'DataZoom': dict({
            'properties': dict({
              'end': dict({
                'default': 100,
                'title': 'End',
                'type': 'integer',
              }),
              'start': dict({
                'default': 0,
                'title': 'Start',
                'type': 'integer',
              }),
              'type': dict({
                'enum': list([
                  'inside',
                  'slider',
                ]),
                'title': 'Type',
                'type': 'string',
              }),
            }),
            'required': list([
              'type',
            ]),
            'title': 'DataZoom',
            'type': 'object',
          }),
          'GridOptions': dict({
            'properties': dict({
              'bottom': dict({
                'anyOf': list([
                  dict({
                    'type': 'string',
                  }),
                  dict({
                    'type': 'integer',
                  }),
                ]),
                'default': '10%',
                'title': 'Bottom',
              }),
              'containLabel': dict({
                'default': True,
                'title': 'Containlabel',
                'type': 'boolean',
              }),
              'left': dict({
                'anyOf': list([
                  dict({
                    'type': 'string',
                  }),
                  dict({
                    'type': 'integer',
                  }),
                ]),
                'default': '10%',
                'title': 'Left',
              }),
              'right': dict({
                'anyOf': list([
                  dict({
                    'type': 'string',
                  }),
                  dict({
                    'type': 'integer',
                  }),
                ]),
                'default': '10%',
                'title': 'Right',
              }),
              'top': dict({
                'anyOf': list([
                  dict({
                    'type': 'string',
                  }),
                  dict({
                    'type': 'integer',
                  }),
                ]),
                'default': '10%',
                'title': 'Top',
              }),
            }),
            'title': 'GridOptions',
            'type': 'object',
          }),
          'ItemStyle': dict({
            'properties': dict({
              'color': dict({
                'anyOf': list([
                  dict({
                    'type': 'string',
                  }),
                  dict({
                    'type': 'null',
                  }),
                ]),
                'default': None,
                'title': 'Color',
              }),
            }),
            'title': 'ItemStyle',
            'type': 'object',
          }),
          'Series': dict({
            'properties': dict({
              'encode': dict({
                '$ref': '#/$defs/SeriesEncode',
              }),
              'itemStyle': dict({
                'anyOf': list([
                  dict({
                    '$ref': '#/$defs/ItemStyle',
                  }),
                  dict({
                    'type': 'null',
                  }),
                ]),
                'default': None,
              }),
              'label': dict({
                'anyOf': list([
                  dict({
                    '$ref': '#/$defs/SeriesLabel',
                  }),
                  dict({
                    'type': 'null',
                  }),
                ]),
                'default': None,
              }),
              'type': dict({
                'enum': list([
                  'bar',
                  'line',
                  'pie',
                  'scatter',
                ]),
                'title': 'Type',
                'type': 'string',
              }),
            }),
            'required': list([
              'type',
              'encode',
            ]),
            'title': 'Series',
            'type': 'object',
          }),
          'SeriesEncode': dict({
            'properties': dict({
              'itemName': dict({
                'anyOf': list([
                  dict({
                    'type': 'string',
                  }),
                  dict({
                    'type': 'null',
                  }),
                ]),
                'default': None,
                'title': 'Itemname',
              }),
              'value': dict({
                'anyOf': list([
                  dict({
                    'type': 'string',
                  }),
                  dict({
                    'type': 'null',
                  }),
                ]),
                'default': None,
                'title': 'Value',
              }),
              'x': dict({
                'anyOf': list([
                  dict({
                    'type': 'string',
                  }),
                  dict({
                    'type': 'null',
                  }),
                ]),
                'default': None,
                'title': 'X',
              }),
              'y': dict({
                'anyOf': list([
                  dict({
                    'type': 'string',
                  }),
                  dict({
                    'type': 'null',
                  }),
                ]),
                'default': None,
                'title': 'Y',
              }),
            }),
            'title': 'SeriesEncode',
            'type': 'object',
          }),
          'SeriesLabel': dict({
            'properties': dict({
              'formatter': dict({
                'anyOf': list([
                  dict({
                    'type': 'string',
                  }),
                  dict({
                    'type': 'null',
                  }),
                ]),
                'default': None,
                'title': 'Formatter',
              }),
              'position': dict({
                'default': 'right',
                'title': 'Position',
                'type': 'string',
              }),
              'show': dict({
                'default': False,
                'title': 'Show',
                'type': 'boolean',
              }),
            }),
            'title': 'SeriesLabel',
            'type': 'object',
          }),
          'TitleOptions': dict({
            'properties': dict({
              'left': dict({
                'anyOf': list([
                  dict({
                    'type': 'string',
                  }),
                  dict({
                    'enum': list([
                      'center',
                      'left',
                      'right',
                    ]),
                    'type': 'string',
                  }),
                ]),
                'default': 'center',
                'title': 'Left',
              }),
              'text': dict({
                'title': 'Text',
                'type': 'string',
              }),
            }),
            'required': list([
              'text',
            ]),
            'title': 'TitleOptions',
            'type': 'object',
          }),
          'Tooltip': dict({
            'properties': dict({
              'formatter': dict({
                'anyOf': list([
                  dict({
                    'type': 'string',
                  }),
                  dict({
                    'type': 'null',
                  }),
                ]),
                'default': None,
                'title': 'Formatter',
              }),
              'trigger': dict({
                'default': 'item',
                'enum': list([
                  'item',
                  'axis',
                ]),
                'title': 'Trigger',
                'type': 'string',
              }),
            }),
            'title': 'Tooltip',
            'type': 'object',
          }),
        }),
        'properties': dict({
          'chart_options': dict({
            '$ref': '#/$defs/ChartOptions',
            'description': 'The options of the chart. A dict, not a json dump',
          }),
          'query_id': dict({
            'description': 'The uuid of the query to execute',
            'title': 'Query Id',
            'type': 'string',
          }),
        }),
        'required': list([
          'chart_options',
          'query_id',
        ]),
        'title': 'Input for `preview_render`',
        'type': 'object',
      }),
    }),
    dict({
      'description': 'Creates a new entity in the catalog.',
      'name': 'SemanticCatalog-semantic_catalog__create_entity',
      'parameters': dict({
        'properties': dict({
          'definition': dict({
            'description': 'The definition of the entity.',
            'title': 'Definition',
            'type': 'string',
          }),
          'entity_name': dict({
            'description': 'The name of the entity to create.',
            'title': 'Entity Name',
            'type': 'string',
          }),
        }),
        'required': list([
          'entity_name',
          'definition',
        ]),
        'title': 'Input for `create_entity`',
        'type': 'object',
      }),
    }),
    dict({
      'description': 'Creates a new issue for an entity.',
      'name': 'SemanticCatalog-semantic_catalog__create_issue',
      'parameters': dict({
        'properties': dict({
          'business_entity_id': dict({
            'description': 'The uuid of the entity to create the issue for.',
            'title': 'Business Entity Id',
            'type': 'string',
          }),
          'description': dict({
            'description': "The description of the issue.                You can use markdown.                Give context / table(s) & column(s) / example(s) to help user visualize the issue.                Use the syntax <QUERY:QUERY_ID> to insert a link to a query if that helps.                Explain the severity of the issue if it's high, critical or blocker.",
            'title': 'Description',
            'type': 'string',
          }),
          'scope': dict({
            'description': 'The scope of the issue ("DATA", "BUSINESS", "BOTH", "UNKNOWN"). Data is for pipeline / warehouse, that can be fixed by data engineers with DBT. Business is for operations / process that impact the business.',
            'title': 'Scope',
            'type': 'string',
          }),
          'severity': dict({
            'description': 'The severity of the issue ("LOW", "MEDIUM", "HIGH", "CRITICAL").',
            'title': 'Severity',
            'type': 'string',
          }),
          'title': dict({
            'description': 'The title of the issue. Tell it as a recommendation.',
            'title': 'Title',
            'type': 'string',
          }),
        }),
        'required': list([
          'title',
          'description',
          'severity',
          'scope',
          'business_entity_id',
        ]),
        'title': 'Input for `create_issue`',
        'type': 'object',
      }),
    }),
    dict({
      'description': 'Deletes an entity from the catalog.',
      'name': 'SemanticCatalog-semantic_catalog__delete_entity',
      'parameters': dict({
        'properties': dict({
          'entity_name': dict({
            'description': 'The name of the entity to delete.',
            'title': 'Entity Name',
            'type': 'string',
          }),
        }),
        'required': list([
          'entity_name',
        ]),
        'title': 'Input for `delete_entity`',
        'type': 'object',
      }),
    }),
    dict({
      'description': 'Fetch an issue by id',
      'name': 'SemanticCatalog-semantic_catalog__read_issue',
      'parameters': dict({
        'properties': dict({
          'issue_id': dict({
            'title': 'Issue Id',
            'type': 'string',
          }),
        }),
        'required': list([
          'issue_id',
        ]),
        'title': 'Input for `read_issue`',
        'type': 'object',
      }),
    }),
    dict({
      'description': 'Updates the quality of an existing entity.',
      'name': 'SemanticCatalog-semantic_catalog__update_entity',
      'parameters': dict({
        'properties': dict({
          'completeness': dict({
            'description': 'The completeness of the entity.',
            'title': 'Completeness',
            'type': 'integer',
          }),
          'definition': dict({
            'description': 'The definition of the entity.',
            'title': 'Definition',
            'type': 'string',
          }),
          'entity_name': dict({
            'description': 'The name of the entity to update.',
            'title': 'Entity Name',
            'type': 'string',
          }),
          'quality_score': dict({
            'description': 'The quality score of the entity.',
            'title': 'Quality Score',
            'type': 'integer',
          }),
          'report': dict({
            'description': 'The quality report of the entity.',
            'title': 'Report',
            'type': 'string',
          }),
          'table_ref': dict({
            'description': 'The table reference of the entity (e.g. "table_name").',
            'title': 'Table Ref',
            'type': 'string',
          }),
        }),
        'required': list([
          'entity_name',
          'definition',
          'completeness',
          'quality_score',
          'report',
          'table_ref',
        ]),
        'title': 'Input for `update_entity`',
        'type': 'object',
      }),
    }),
    dict({
      'description': 'Updates an existing issue.',
      'name': 'SemanticCatalog-semantic_catalog__update_issue',
      'parameters': dict({
        'properties': dict({
          'description': dict({
            'description': 'The description of the issue.',
            'title': 'Description',
            'type': 'string',
          }),
          'issue_id': dict({
            'description': 'The id of the issue to update.',
            'title': 'Issue Id',
            'type': 'string',
          }),
          'scope': dict({
            'description': 'The scope of the issue ("DATA", "BUSINESS", "BOTH", "UNKNOWN"). Data is for pipeline / warehouse, that can be fixed by data engineers with DBT. Business is for operations / process that impact the business.',
            'title': 'Scope',
            'type': 'string',
          }),
          'severity': dict({
            'description': 'The severity of the issue ("LOW", "MEDIUM", "HIGH", "CRITICAL").',
            'title': 'Severity',
            'type': 'string',
          }),
          'status': dict({
            'description': 'The status of the issue ("OPEN", "IN_PROGRESS", "DONE").',
            'title': 'Status',
            'type': 'string',
          }),
          'title': dict({
            'description': 'The title of the issue.',
            'title': 'Title',
            'type': 'string',
          }),
        }),
        'required': list([
          'issue_id',
          'status',
          'title',
          'description',
          'scope',
          'severity',
        ]),
        'title': 'Input for `update_issue`',
        'type': 'object',
      }),
    }),
    dict({
      'description': 'Fetch all model details from the DBT catalog.',
      'name': 'DBT-dbt__fetch_model',
      'parameters': dict({
        'properties': dict({
          'key': dict({
            'title': 'Key',
            'type': 'string',
          }),
        }),
        'required': list([
          'key',
        ]),
        'title': 'Input for `fetch_model`',
        'type': 'object',
      }),
    }),
    dict({
      'description': 'Return a list of models in the DBT catalog. Return: key, description',
      'name': 'DBT-dbt__fetch_model_list',
      'parameters': dict({
        'properties': dict({
        }),
        'title': 'Input for `fetch_model_list`',
        'type': 'object',
      }),
    }),
    dict({
      'description': 'Search for a model in the DBT catalog. Return: key, description',
      'name': 'DBT-dbt__search_models',
      'parameters': dict({
        'properties': dict({
          'query': dict({
            'title': 'Query',
            'type': 'string',
          }),
        }),
        'required': list([
          'query',
        ]),
        'title': 'Input for `search_models`',
        'type': 'object',
      }),
    }),
    dict({
      'description': None,
      'name': 'Notes-notes__close_note',
      'parameters': dict({
        'properties': dict({
          'id': dict({
            'title': 'Id',
            'type': 'integer',
          }),
        }),
        'required': list([
          'id',
        ]),
        'title': 'Input for `close_note`',
        'type': 'object',
      }),
    }),
    dict({
      'description': None,
      'name': 'Notes-notes__create_note',
      'parameters': dict({
        'properties': dict({
          'content': dict({
            'title': 'Content',
            'type': 'string',
          }),
          'title': dict({
            'title': 'Title',
            'type': 'string',
          }),
        }),
        'required': list([
          'title',
          'content',
        ]),
        'title': 'Input for `create_note`',
        'type': 'object',
      }),
    }),
    dict({
      'description': 'Delete a note from the project',
      'name': 'Notes-notes__delete_note',
      'parameters': dict({
        'properties': dict({
          'id': dict({
            'title': 'Id',
            'type': 'string',
          }),
        }),
        'required': list([
          'id',
        ]),
        'title': 'Input for `delete_note`',
        'type': 'object',
      }),
    }),
    dict({
      'description': None,
      'name': 'Notes-notes__list_notes',
      'parameters': dict({
        'properties': dict({
        }),
        'title': 'Input for `list_notes`',
        'type': 'object',
      }),
    }),
    dict({
      'description': None,
      'name': 'Notes-notes__open_note',
      'parameters': dict({
        'properties': dict({
          'id': dict({
            'title': 'Id',
            'type': 'integer',
          }),
        }),
        'required': list([
          'id',
        ]),
        'title': 'Input for `open_note`',
        'type': 'object',
      }),
    }),
  ])
# ---
# name: test_tools
  dict({
    'database': 'DatabaseTool',
    'dbt': 'DBT',
    'echarts': 'EchartsTool',
    'notes': 'Notes',
    'semantic_catalog': 'SemanticCatalog',
    'workspace': 'WorkspaceTool',
  })
# ---
